### DISCUSSION TITLE
[V5] Eventless (transient) transition syntax

### DISCUSSION THREAD

**Discussion by davidkpiano** [statelyai/xstate#1243] [General] [answered]
In V4, to define what SCXML calls an **eventless transition** (which is a `<transition>` without an event, or triggered by a "null event") is defined like this:

```js
someState: {
  on: {
    '': {
      target: 'otherState',
      cond: 'someCond'
    },
    CLICK: 'clicked'
  }
}
```

However, that syntax can be confusing, is not ideal for searchability, and can "feel" like a code-smell (or design-smell), even though it isn't.

There were some options we've talked about for defining a _separate property_ for eventless transitions:

- `onAuto` (there are a few places where this is called an "automatic transition", although that can be confusing)
  - https://statecharts.github.io/glossary/automatic-transition.html
  - https://www.codeproject.com/Articles/43356/A-Simple-State-Machine
  - http://www.patrickvanbergen.com/on_fsm/on_fsm.html
- ~`onNull`~ (quite literal)
- `transient` (this term isn't often used outside of XState)
- `immediate` or `immediately` (conveys that the transition is immediately taken)
- **`always`**, which is [what Yakindu calls the trigger](https://www.itemis.com/en/yakindu/state-machine/documentation/user-guide/quick_ref_states_transitions#quick_ref_states_transitions), and it makes sense:

```js
someState: {
  // Always transition to `otherState` immediately if the `cond` is met
  always: {
    target: 'otherState',
    cond: 'someCond'
  },
  on: {
    CLICK: 'clicked'
  }
}
```

Any thoughts? I would like to come to a consensus on an intuitive name for this for V5.

**Comment by Andarist** [comment, marked as answer]
I don't find `''` that much confusing (visually) or feeling like a code-smell, but a searchability argument is somewhat compelling - although if you include a trailing `:` you shouldn't really bump into false positives.

A separate thing is that if `''` alone is clear in what it does - when I've started working with XState it definitely was not clear for me that this would be checked **always**. I think I've learned about this reading through SCXML spec and it was a surprise to me - even if it was mentioned in the XState docs anywhere, but you know how it is, we usually just skim through docs 😅 For this reason, I find `always` the best out of proposed alternatives and I don't think we could figure out anything better that would be, at the same time, concise.

**Comment by danielkcz** [comment]
The `''` was a big hurdle for me as well, it looked weird at first sight and I didn't know what it means. The bad part is you cannot simply search docs for `'':` and you don't know it's called _transient_.

The `always` seems indeed nice, but shouldn't it be prefixed with `on`? I mean there is `entry/exit` which are actions. Wouldn't be confusing to have `always` thinking like _these are actions that executed always_? I think that unless you want to drop `onDone/onError` then `onAlways` would be better to distinguish that it's a transition.

**Comment by Andarist** [comment]
`onDone`/`onError` are triggered by actual events, whereas `always` is not. `_event` in transient transitions should be bound to the last received event. So in a sense, this doesn't actually work like `on*`/`on` as those are matched with real events that are always sent to a machine explicitly from other entities.

**Reply by davidkpiano** [reply]
Yes, and also, we have `after`, which is not prefixed with `onAfter`.
